odoo/service/shared_memory.py

```diff
@@ -0,0 +1,446 @@
+
+    def _del_index(self, index, entry):
+        """
+        Remove the entry at `index` and data linked to.
+        It compacts the hash table to ensure the correctness
+        of the _lookup (linear probing)
+
+        ! Need lock !
+        """
+        if entry.prev == entry.next == index:  # If am the only one
+            self._root = -1
+        else:
+            self._entry_table[entry.next].prev = entry.prev
+            self._entry_table[entry.prev].next = entry.next
+            if self._root == index:
+                self._root = entry.next
+
+        self._entry_table[index] = (0, -1, -1)
+        self._length -= 1
+        free_size = self._free(index)
+
+        # Delete the keys that are between this element, and the next free spot, having
+        # an index lower or equal to the position we delete (conflicts handling).
+        def move_index(old_index, new_index):
+            old_entry = self._entry_table[old_index]
+            hash_i, prev_i, next_i = old_entry.hash, old_entry.prev, old_entry.next
+            if self._entry_table[old_index].next == old_index:  # if it is the only one item
+                prev_i = next_i = new_index
+            self._entry_table[prev_i].next = self._entry_table[next_i].prev = new_index
+            self._entry_table[new_index] = (hash_i, prev_i, next_i)
+            self._entry_table[old_index] = (0, -1, -1)
+            self._data_idx[new_index] = self._data_idx[old_index]
+            self._data_idx[old_index] = (-1, -1)
+            if self._root == old_index:
+                self._root = new_index
+
+        index_empty = index
+        for i in range(index + 1, index + self._size):
+            i_mask = i % self._size  # from index -> self._size -> 0 -> index - 1
+            i_entry = self._entry_table[i_mask]
+            if i_entry.prev == -1:
+                break
+            hash_mask = i_entry.hash % self._size
+
+            distance_i = i_mask - hash_mask if i_mask >= hash_mask else self._size - hash_mask + i_mask
+            # distance error of i,
+            # - if 0 then he is a the correct location don't move
+            # - if < than distance_new = not suitable location
+            # else compress
+            if distance_i == 0:
+                continue
+            distance_new = index_empty - hash_mask if index_empty >= hash_mask else self._size - hash_mask + index_empty
+            if distance_new < distance_i:
+                move_index(i_mask, index_empty)
+                index_empty = i_mask
+        else:
+            raise MemoryError("The hashtable seems full, it doesn't make any sense")
+
+        return free_size
+
+    def _data_get(self, index):
+        """
+        Get (key, value) of entry at `index`.
+
+        ! Need lock !
+        """
+        data_id = self._data_idx[index]
+        return marshal.loads(self._data[data_id.position:data_id.position + data_id.size])
```

Comment:
Transformations (marshal dumps/loads) of data must be done outside the share memory so that the lock takes the least amount of time.

Reply:
Just for the record:
I did it, but it becomes more complex for the data structure (need the size of the key) and you need to copy the data of value before release the lock to avoid any incoherence (with bytes). 
With these change, in a strong concurrent situation and with data simulating t-cache, it wasn't faster. And in small concurrent situation (or in mono-process), it was slower (more than 10 %).